py/picca/io.py

```diff
@@ -262,51 +277,111 @@ def read_from_pix(in_dir,pix,thid,ra,dec,zqso,plate,mjd,fid,order,log=None):
+                if card in head:
+                    nexp = head["NEXP_{}".format(c)]
+                else:
+                    continue
+                for i in range(nexp):
+                    str_iexp = str(iexp)
+                    if iexp<10:
+                        str_iexp = '0'+str_iexp
+                    
+                    card = "EXPID"+str_iexp
+                    if not card in head:
+                        continue
+
+                    exps.append(head["EXPID"+str_iexp][:11])
+                    iexp += 1
+
+        print "INFO: found {} exposures in plate {}".format(len(exps), p)
+    
+        if len(exps) == 0:
+            continue
 
-        name = in_dir+"/{}/{}-*-*.fits{}".format(p,prefix,sufix)
-        fi = glob.glob(name)
-        w = (plate==p)
+        exp_num = [e[3:] for e in exps]
+        exp_num = sp.unique(exp_num)
+        sp.random.shuffle(exp_num)
+        exp_num = exp_num[0]
+        for exp in exps:
+            if single_exp:
+                if not exp_num in exp:
+                    continue
+            t0 = time.time()
+            ## find the spectrograph number:
+            spectro = int(exp[1])
+            assert spectro == 1 or spectro == 2
 
-        if len(fi)==0:
-            sys.stderr.write("No files found in {}".format(name))
-            continue
+            ## find out the fibers where the qsos are:
+            if spectro == 1:
+                wfib = wplate & (fid <= 500)
+            if spectro == 2:
+                wfib = wplate & (fid > 500)
 
-        for the_file in fi:
+            if wfib.sum()==0:
+                continue
 
-            sys.stderr.write("reading {}\n".format(the_file))
-            h = fitsio.FITS(the_file)
-            fib_list=list(h[5]["FIBERID"][:])
-            flux = h[0].read()
-            ivar = h[1].read()*(h[2].read()==0)
-            llam = h[3].read()
+            plate_fibs = fid[wfib]
 
-            if "-r1-" in the_file or "-b1-" in the_file:
-                ww = w & (fid<=500)
-            elif "-r2-" in the_file or "-b2-" in the_file:
-                ww = w & (fid>=501)
+            ## collect the relevant flux, ivar and ll in a list:
+            flux = []
+            ivar = []
+            llam = []
 
-            for (t, r, d, z, p, m, f) in zip(thid[ww], ra[ww], dec[ww], zqso[ww], plate[ww], mjd[ww], fid[ww]):
+            spcframe = fitsio.FITS(in_dir+"/{}/spCFrame-{}.fits".format(p, exp))
 
-                index = fib_list.index(f)
+            flux = spcframe[0].read()
+            ivar = spcframe[1].read()*(spcframe[2].read()==0)
+            llam = spcframe[3].read()
+            
+            ## now convert all those fluxes into forest objects
+            for index, (t, r, d, z, p, m, f) in enumerate(zip(thid[wfib], ra[wfib], dec[wfib], zqso[wfib], plate[wfib], mjd[wfib], fid[wfib])):
+                index =(f-1)%500
+                wlam = (10**llam[index]/(1+z) > 1220) & (10**llam[index]/(1+z)<1230)
```

Comment:
why 1220 and 1230? and I don't see where you use this.

Reply:
woops, leftover from debugging...

Reply:
removed